java/org/contikios/cooja/Simulation.java

```diff
@@ -152,68 +156,117 @@ public Simulation(Cooja cooja, long seed) {
     this.cooja = cooja;
     randomGenerator = new SafeRandom(this);
     randomSeed = seed;
+    simulationThread = new Thread(() -> {
+      boolean isAlive = true;
+      do {
+        boolean isSimulationRunning = false;
+        EventQueue.Pair nextEvent = null;
+        try {
+          while (isAlive) {
+            Object cmd;
+            do {
+              cmd = isSimulationRunning ? commandQueue.poll() : commandQueue.take();
+              if (cmd instanceof Runnable r) {
+                r.run();
+              } else if (cmd instanceof Command c) {
+                isAlive = c != Command.QUIT;
+                isShutdown = !isAlive;
+                isSimulationRunning = c == Command.START;
+                setRunning(isSimulationRunning);
+              }
+            } while (cmd != null && isAlive);
+
+            if (isSimulationRunning) {
+              /* Handle one simulation event, and update simulation time */
+              nextEvent = eventQueue.popFirst();
+              assert nextEvent != null : "Ran out of events in eventQueue";
+              assert nextEvent.time >= currentSimulationTime : "Event from the past";
+              currentSimulationTime = nextEvent.time;
+              nextEvent.event.execute(currentSimulationTime);
+            }
+          }
+        } catch (RuntimeException e) {
+          if ("MSPSim requested simulation stop".equals(e.getMessage())) {
+            /* XXX Should be*/
+            logger.info("Simulation stopped due to MSPSim breakpoint");
+          } else {
+            logger.fatal("Simulation stopped due to error: " + e.getMessage(), e);
+            if (!Cooja.isVisualized()) {
+              /* Quit simulator if in test mode */
+              System.exit(1);
+            } else {
+              String title = "Simulation error";
+              if (nextEvent != null && nextEvent.event instanceof MoteTimeEvent moteTimeEvent) {
+                title += ": " + moteTimeEvent.getMote();
+              }
+              Cooja.showErrorDialog(title, e, false);
+            }
+          }
+        } catch (InterruptedException e) {
+          // Simulation thread interrupted - quit
+          logger.warn("simulation thread interrupted");
+          isAlive = false;
+          isShutdown = true;
+        }
+        setRunning(false);
```

Comment:
Yes, I think so. For example, if the simulation is stopped by breakpoint or by exception.

Reply:
I think you're right. Shouldn't the catch-blocks be combined into `} catch (InterruptedException | RuntimeException e) { .. }`, with the `isAlive = ..` inside the else-branch on line 202 in that case?

Reply:
And second question: can the call to `setRunning()` be replaced with `commandQueue.putFirst(Command.STOP/QUIT)` and letting the regular loop shut down the simulation (assuming commandQueue is converted to a ConcurrentLinkedDequeue)?

Reply:
They are doing quite different things. Interrupting the thread will cause the simulation to shut down, but other exceptions will only stop the simulation, possibly with a warning, and perhaps exit Cooja if running without GUI. I can combine the code blocks, but I do not think they have that much in common. Also, I tried to leave the old exception handling code as unchanged as possible in this PR.

Reply:
For the second question: No, I do not think so. If the simulation thread has been interrupted - no more commands will be processed.

Reply:
I should stop sketching on details when I ask structural questions, the only implementation of the BlockingDeque interface is LinkedBlockingDeque. My question was basically if inserting the stop/quit command first in the queue would allow re-use of the regular shutdown mechanism instead of having to call `setRunning(false)`.

Reply:
You're probably right, there's a value of getting this landed, as long as it doesn't look terrible it can be simplified in a future PR.